olive/auto_optimizer/__init__.py

```diff
@@ -0,0 +1,232 @@
+        self.regulate_fp16(pass_config, pass_flows)
+        self.regulate_data_config(pass_config, pass_flows)
+        return pass_config, pass_flows
+
+    def suggest_passes_from_template(self):
+        from olive.auto_optimizer.template_mapping import get_model_template_config, get_model_template_passes
+
+        assert self.auto_optimizer_config.opt_level == 0, "opt_level must be 0 for suggest_pass_flows_from_template"
+
+        pass_flows = get_model_template_passes(self.model_type)
+        pass_config = get_model_template_config(self.model_type)
+
+        for pass_item in pass_flows:
+            for p_name in pass_item:
+                user_defined_config = self.auto_optimizer_config.customized_pass_config.get(p_name, {})
+                if p_name not in pass_config:
+                    pass_config[p_name] = user_defined_config
+                else:
+                    pass_config[p_name].update(user_defined_config)
+        # if there are duplicated passes in template, we should raise error
+        assert all(len(p) == len(set(p)) for p in pass_flows), f"duplicated passes in template: {pass_flows}"
+        return pass_config, pass_flows
+
+    def regulate_fp16(self, pass_config, pass_flows):
+        if not self.is_gpu or not self.is_accuracy_drop_tolerance:
+            return
+
+        is_cuda_ep = self.accelerator_spec.execution_provider == "CUDAExecutionProvider"
+        is_trt_ep = self.accelerator_spec.execution_provider == "TensorrtExecutionProvider"
+        assert (
+            is_cuda_ep or is_trt_ep
+        ), "can not support CUDAExecutionProvider and TensorrtExecutionProvider at the same time"
+
+        customized_fp16 = self.allow_precision(Precision.FP16)
+        cuda_fp16 = customized_fp16 and is_cuda_ep
+        trt_fp16 = customized_fp16 and not cuda_fp16
+
+        if "OrtTransformersOptimization" in pass_config:
+            # if cuda ep, enable fp16 for transformers optimization but not for tensorrt
+            self.set_config_value_if_not_exist(pass_config["OrtTransformersOptimization"], "float16", cuda_fp16)
+            self.set_config_value_if_not_exist(pass_config["OrtTransformersOptimization"], "use_gpu", True)
+            if "OrtPerfTuning" in pass_config:
+                # if trt ep, disable fp16 for transformers optimization but enable fp16 for perf_tuning
+                self.set_config_value_if_not_exist(pass_config["OrtPerfTuning"], "trt_fp16_enable", trt_fp16)
+                # enable_cuda_graph only for cuda ep
+                self.set_config_value_if_not_exist(pass_config["OrtPerfTuning"], "enable_cuda_graph", cuda_fp16)
+                self.set_config_value_if_not_exist(pass_config["OrtPerfTuning"], "io_bind", True)
+
+    def regulate_passes(self, pass_config, pass_flows):
+        # remove passes that are not supported by execution provider
+        to_del_passes = set(self.auto_optimizer_config.disabled_passes) or set()
```

Comment:
This should be the responsibility of the Pass itself, not the auto_optimizer or engine.
Imagine a Pass contributed by external developer for external target.

Reply:
This part is a bit different with the config pruning or filtering which are used to cut off invalid configs for a specific given pass.

The `regulate_passes` is about to globally regulate passes combinations, like the cases of :
- quantization and fp16 cannot appear simultaneously
- cuda fp16(transformer optimization) and trt fp16(perf_tuning) cannot appear at the same time.

It can improve the olive optimization/search efficiency with global regulating in advance.